drivers/w3s.go

```diff
@@ -0,0 +1,340 @@
+package drivers
+
+import (
+	"context"
+	"fmt"
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-unixfs"
+	"github.com/ipld/go-car"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+const w3SDefaultSaveTimeout = 5 * time.Minute
+
+var cidV1 = merkledag.V1CidPrefix()
+
+// This represents the main CAR directory structure organized by pubId.
+// Data for each pubId is removed after the CAR directory is published.
+var (
+	dataToPublish   = make(map[string]*rootCar)
```

Comment:
Yes, exactly. Technically, you're right, it should be just a field in the driver or session, but unfortunately it's not how it's used right now. Currently, every time you interact with the storage, you create a new driver and a new session. I think we can (and should) change it in the future, but I don't want to touch it right now, since I think these will cause quite a lot of changes in the catalyst-api (and possibly other projects).